### PR TITLE
feat(new transform): Inital `kubernetes_pod_metadata` transform

### LINKED ISSUE #1072: New `kubernetes_pod_metadata` transform
We should provide a `kubernetes_pod_metadata` transform that is able to fetch a list of pods and enrich events. This should work based off of the kube openapi spec using the [`k8-openapi`] crate.

# Configuration

```toml
[transforms.my_transform_id]
  type = "kubernetes_pod_metadata"
  fields = ["instance-id", "local-hostname", "local-ipv4", "public-hostname", "public-ipv4"] # default
  namespace = ["your-namespace"] # optional, default is to list ALL pods in all namespaces
```

## Options

- `fields` contains the list of fields to include in each event.
- ~~`namespace` contains a list of namespaces that we should watch for a list of pods and their metadata.~~ See global schema for this field.

## Fields

We should provide this set of fields by default and they are overridable via the `fields` option.

- `pod_name`
- `pod_namespace`
- `pod_uid`
- `labels` (not sure if we should provide a way to include the label name in the key or have them as a tuple?)
- `annotations`
- `node_name`

Most of these data points should come from a combination of [`PodSpec`] and [`ObjectMeta`] using the [`Pod::list_namespaced_pod`].

[`k8-openapi`]: https://docs.rs/k8s-openapi
[`PodSpec`]: https://docs.rs/k8s-openapi/0.6.0/k8s_openapi/api/core/v1/struct.PodSpec.html
[`ObjectMeta`]: https://docs.rs/k8s-openapi/0.6.0/k8s_openapi/apimachinery/pkg/apis/meta/v1/struct.ObjectMeta.html
[`Pod::list_namespaced_pod`]: https://docs.rs/k8s-openapi/0.6.0/k8s_openapi/api/core/v1/struct.Pod.html#method.list_namespaced_pod

## Global Schema

`log_schema` should add a `kubernetes_namespace` field to correlate where nested fields should be added to events. This should replace the `namespace` transform option.

## Implementation

This should use the [`k8-openapi`] crate, plus the [`kube`] crate, though I would prefer we stick with just hyper and vendor the config loading. This implementation should follow the proposed method for implementing a stream table join transforms as seen in #1069.

[`kube`]: https://docs.rs/kube/0.16.1/kube/

## Questions

- How should we nest these? Should they live under a `kubernetes` key?
- How can we correlate pod events coming from the `kubernetes` source with the ones on the list? @ktff 

## Todo

- [x] `WatchClient`

- [ ] `kubernetes_pod_metadata`

### PR DESCRIPTION
**Pull request by ktff**
Closes #1072

### Changes:
 * `namespace` field isn't implemented as it isn't needed. [More info](https://github.com/timberio/vector/issues/1072#issuecomment-574258208). 

### Implemetation notes:
 * We have only one request per Node on `/api/v1/pods` endpoint through which we are initializing and watching metadata for all Pods on Node.
 * Implemented with `k8-openapi` and hyper.
 * Supported fields:
   * `name`
   * `namespace`
   * `creation_timestamp`
   * `deletion_timestamp`
   * `labels`
   * `annotations`
   * `node_name`
   * `hostname`
   * `priority`
   * `priority_class_name`
   * `service_account_name`
   * `subdomain`
   * `host_ip`
   * `ip`
 * Fields are added to `Events` with `pod_` prefix. example. `pod_labels`. I think `pod` better describes what are the fields describing then `kubernetes`. And prefix is used, instead of nested key, to have same naming scheme as `kubernetes_source`.  
 * Tests and example config are using RBAC authorization to allow Vector access to API server. RBAC is usualy enabled by default.
 * Tested on GKE 1.13.11-gke.23 and AWS EKS 1.14
 * Uses Kubernetes API v1.15

### Todo
 

- [ ]  After this is merged, open a separate PR to rename `kubernetes_source` `object_uid` and `pod_uid` fields. This will be a breaking change.

### CODE REVIEW

File: src/transforms/kubernetes_pod_metadata.rs [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+    async fn watch(&mut self, mut version: Option<Version>, request: Request<Body>) -> WatchResult {
+        // Start watching
+        let response = self.client.request(request).compat().await;
+        match response {
+            Ok(response) => {
+                info!(message = "Watching Pod list for changes.");
+                let status = response.status();
+                let mut unused = Vec::new();
+                let mut body = response.into_body();
+                'watch: loop {
```
Comment (LucioFranco) [inline src/transforms/kubernetes_pod_metadata.rs]: this is a bit hard to read, is there a way we can break this up a bit? Generally, I am not a fan of large socpes with named breaks. This code also kinda drifts to the right. Happy to think through this with you but I'd like to see if we can improve this.

File: src/transforms/kubernetes_pod_metadata.rs [hunk 2]
```diff
@@ -0,0 +1,767 @@
...
+    #[snafu(display("Uri gotten from Kubernetes is invalid: {}.", source))]
+    InvalidUriParts { source: uri::InvalidUriParts },
+    #[snafu(display("Authorization token gotten from Kubernetes is invalid: {}.", source))]
+    InvalidToken { source: header::InvalidHeaderValue },
+}
+
+/// Background client which watches for Pod metadata changes and propagates them to Transform.
+struct MetadataClient {
+    fields: Vec<Box<dyn Fn(&Pod) -> Vec<(Atom, Value)> + Send + Sync + 'static>>,
+    sender: Option<mpsc::Sender<(Bytes, Vec<(Atom, Value)>)>>,
```
Comment (LucioFranco) [inline src/transforms/kubernetes_pod_metadata.rs]: Why a channel over using `evmap`?

**Comment by LucioFranco** [comment]
I left you a few high level questions, this change is decently large so I'd like to see if we can maybe split up parts of it or at least try to reduce somethings down. My review queue is a bit backed up right now so I'll give this a bit more of a thorough review soon.

File: src/transforms/kubernetes_pod_metadata.rs (hunk 2 above)
Comment (ktff) [reply src/transforms/kubernetes_pod_metadata.rs]: Didn't want to depend on extra dependencies. 

But as we are already depending on `evmap`, it's a better option, so I'll change it to that.

**Comment by binarylogic** [comment]
@ktff this looks nice! I'm excited to get this merged. Given that this is mostly done, is there any way we can break this up into smaller changes and submit them as separate pull requests? I don't want to create more work for you, I also don't understand the code enough to know if this is even possible. But this will help to ensure we provide high-quality reviews. For example, maybe we can submit the metadata client separately since that looks rather large? I assume you'll know better ways to break this up. Let me know what you think.

**Comment by ktff** [comment]
@LucioFranco 
> split up parts of it or at least try to reduce somethings down

Good question, we could in advance split out `MetadataClient` and make it a generic Kubernetes `WatchClient`. Then we could use it for this transform and #1293.

Such `WatchClient` would have ~400 lines, and transform would have ~300 lines.

**Comment by ktff** [comment]
@binarylogic I was just writing about it :)

**Comment by binarylogic** [comment]
No problem! And that sounds good. If you agree, why don't you close this one and open new PRs?

**Comment by ktff** [comment]
@binarylogic Will do.

**Comment by binarylogic** [comment]
@ktff do you plan on opening follow-up PRs? Let us know if we can help any.

**Comment by ktff** [comment]
@binarylogic 

I am planning in submitting `WatchClient`, that is in the works, first and then `transform`. 
Basically everything is written, it's "just" a matter of cleanly separating concerns.

**Pull request closed without merging**